### PR TITLE
Spies

### PR DESCRIPTION
**Pull request by blerner**
This introduces support for `spy`, which looks like

```
spy <expr-title>:
  name,
  name: expr
  ...
end
```

This will render as a spyblock, with the `<expr-title>` as a name for the block.  The contents of the block will render as Each `name: expr` clause will rich-render the expression, and label it with name.  `name` clauses are shorthand for `name: name`, i.e., the identifier is eponymous with its value.

### CODE REVIEW

**Comment by jpolitz** [comment]
I think we should not include 72caa5b1a607081b1040fdcf635f9082c814d9f2, because it will be surprising since we didn't announce it earlier.

**Comment by jpolitz** [comment]
This is a gratifyingly small PR.  I think this is a great first run to try out.

File: src/arr/trove/ast.arr [hunk 1]
```diff
@@ -1166,10 +1166,37 @@ data TableRow:
     method tosource(self):
       PP.flow([list: str-rowcolon,
           PP.flow-map(PP.commabreak, _.tosource(), self.elems)])
+  | s-spy-block(l :: Loc, message :: Option<Expr>, contents :: List<SpyField>) with:
```
Comment (jpolitz) [inline src/arr/trove/ast.arr]: Pretty sure this isn't a TableRow.

**Comment by blerner** [comment]
I really think we should include the 72caa5b change, since there are likely not to be many people yet relying on it, and it'll be easier to change now as we introduce the intended replacement, than if we let it sit another year.

**Comment by shriram** [comment]
PLEASE do NOT mess with how `print` works. I actually use it in class. My students depend on it. (Also, some of my students think it's the coolest thing ever and wonder why other languages don't do the same thing.) I do not want this to ever change in Pyret.

**Comment by shriram** [comment]
Is the <expr-title> part optional? Otherwise this looks awfully heavy-weight.

**Comment by blerner** [comment]
What the...?  We _*never*_ agreed that print should continue to return its value; it was always a hack (you explicitly called it a hack) that was the whole reason we talked about introducing spy in the first place.  This isn't the right venue to rehash the entire design discussion, though.

Yes, the name is optional.

**Comment by shriram** [comment]
It may have been a hack, but I've come to see it as a pretty useful feature. `spy` does not at all replace its value. And I agree, this isn't the place to discuss this. We should just leave `print` alone for now. I see no harm in doing so.

**Comment by schanzer** [comment]
Linking this to #179

**Comment by jpolitz** [comment]
@blerner should we merge this without https://github.com/brownplt/pyret-lang/commit/72caa5b1a607081b1040fdcf635f9082c814d9f2?  It seems worth getting some progress made on it, and we keep wanting it.

File: src/js/base/runtime.js [hunk 2]
```diff
@@ -5049,6 +5050,79 @@ function (Namespace, jsnums, codePoint, seedrandom, util) {
...
+    function toReprArray(vals, reprMethod) {
+      // vals should be an array of {name: String, val: PyretValue, [method: ReprMethod]} objects
+      // return is an array of either {name: String, val: rendered} objects (on success)
+      // or {name: String, exn: exn} objects (on failure)
+      // but all will run.
+      // Each item can specify its own reprmethod, or the default one can be used
+      var results = [];
+      return thisRuntime.safeCall(function() {
+        return thisRuntime.eachLoop(makeFunction(function(i) {
+          thisRuntime.pauseStack(function(restarter) {
```
Comment (jpolitz) [inline src/js/base/runtime.js]: This needs a `return`

**Pull request merged**